---
title: "Iglon5 Differential Expression"
author: "Andrew G. Newman"
output:
  html_document: default
  pdf_document: default
---

# Setup Document and Space

## Code to setup markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE) 
if(!require(formatR))
  install.packages("formatR")
## Added for multi-output code chunks
doc.type <- knitr::opts_knit$get('rmarkdown.pandoc.to')
## Added so code chunks don't go off the page in pdf
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60), tidy=TRUE, message= F, warning = F)
if(knitr:::is_latex_output()) options(knitr.graphics.auto_pdf = TRUE)
```

## Proxy Set up

```{r setproxy}
#SETPROXY ONLY IF ON CHARITE NETWORK
Sys.setenv("http_proxy" = "http://proxy.charite.de:8080") # if on charite network
Sys.setenv("https_proxy" = "http://proxy.charite.de:8080") # if on charite network
# # #UNSET
# Sys.setenv("http_proxy" = "") # if not
# Sys.setenv("https_proxy" = "") # if not
```




# Alignment performed on BIH Cluster

Config file used for Alignment: "/fast/users/newmana_c/config/Iglon5_config.yaml":
```{bash, eval = FALSE}
---
samples:
  PBS_1: P2032_648_LCTX1_PBS_S3
  PBS_2: P2032_658_LCTX1_PBS_S6
  PBS_3: P2032_661_LCTX1_PBS_S9
  PBS_4: P2032_666_LCTX1_PBS_S10
  IgG_1: P2032_617_LCTX1_PatIgG_S1
  IgG_2: P2032_649_LCTX1_PatIgG_S4
  IgG_3: P2032_659_LCTX1_PatIgG_S7
  IgG_4: P2032_667_LCTX1_PatIgG_S11
  Iglon5_1: P2032_618_LCTX1_IgL5_S2
  Iglon5_2: P2032_650_LCTX1_IgL5_S5
  Iglon5_3: P2032_660_LCTX1_IgL5_S8
genomedir: "/fast/users/newmana_c/work/genomes/GRCm39"
star_index: "/fast/users/newmana_c/work/genomes/GRCm39"
TE_gtf_file: "/fast/users/newmana_c/work/genomes/GRCm39/GRCm39_Ensembl_rmsk_TE.gtf"
gene_gtf_file: "/fast/users/newmana_c/work/genomes/GRCm39/Mus_musculus.GRCm39.109.chr.gtf"
gene_gff_file: "/fast/users/newmana_c/work/genomes/GRCm39/gencode.vM31.primary_assembly.basic.annotation_prepared.gff"
workdir: "/fast/users/newmana_c/work/Iglon5"
download_dir: "/fast/users/scratch/Iglon5/P2032"
```

Raw fastq files were aligned to GRCm39 using the following Snakemake pipeline that retains repetitive elements and non-coding RNA.

```{python, python.reticulate = FALSE, eval = FALSE, echo = TRUE, results = FALSE, tidy=TRUE, tidy.opts=list(width.cutoff=60)}

#! bin/env/python
#TETranscripts Pipeline
#Invoked by submit_TE.sh in ~/work/Iglon5

configfile: "/fast/users/newmana_c/config/Iglon5_config.yaml"
#DOWNLOAD_DIR = config["download_dir"]


#Variables from Config
SAMPLES = config["samples"]
TE_GTFFILE = config["TE_gtf_file"]
GTFFILE = config["gene_gtf_file"]
STAR_INDEX = config["star_index"]

#Outputs:
#STAR = expand("alignment/{sample}.sorted.bam", sample = config["samples"])
FLAGSTAT = expand("alignment/{sample}.sorted.bam.flagstat", sample = config["samples"])
ALL_BAM_INDEX = expand("alignment/{sample}.sorted.bam.bai", sample = config["samples"])
QC = expand("FastQC/{sample}_R{num}_fastqc.zip", sample=config["samples"], num=[1,2])
TE_OUT = expand("counts/{sample}.cntTable", sample = config["samples"])
MULTIQC = "FastQC/fastq_multiqc.html"

rule all:
    input: TE_OUT, ALL_BAM_INDEX, FLAGSTAT, MULTIQC


rule fastqc:
    input:
        sample=lambda wildcards: expand(f"fastq/{config['samples'][wildcards.sample]}_L001_R{{num}}_001.fastq.gz", num = [1,2])
    output:
        # Output needs to end in '_fastqc.html' for multiqc to work
        html="FastQC/{sample}_R{num}_fastqc.html",
        zip="FastQC/{sample}_R{num}_fastqc.zip"
    wrapper:
        "0.21.0/bio/fastqc"


rule multiqc_fastq:
    input:
        expand("FastQC/{sample}_R{num}_fastqc.html", sample=config["samples"],num=[1,2])
    output:"FastQC/fastq_multiqc.html"
    resources: 
        mem='3G',
        time='04:00:00'
    log: "logs/multiqc.log"
    shell:
        """
        multiqc FastQC \
        -n {output} 
        """

rule flagstat_bam:
    input:
        "alignment/{sample}.sorted.bam"
    output:
                "alignment/{sample}.sorted.bam.flagstat"
    log:
        "logs/{sample}.flagstat_bam"
    shell:
        """
        samtools flagstat {input} > {output} 2> {log}
        """

rule index_bam:
	input:
		"alignment/{sample}.sorted.bam"
	output:
		"alignment/{sample}.sorted.bam.bai"
	log:
                "logs/{sample}.index_bam"
	threads: 1
        resources:
            mem='10G',
            time='04:00:00'
	shell:
                """
                samtools index {input} 2> {log}
                """
rule TET_all_counts:
        input:
                gtf = GTFFILE,
                te = TE_GTFFILE,
                BAI = ALL_BAM_INDEX,
                bam = "alignment/{sample}.sorted.bam"
        output: "counts/{sample}.cntTable"
        threads: 8
        resources:
            mem_mb=24000
        shell:"""
              TEcount \
              --mode multi \
              --format BAM \
              --sortByPos \
              -b {input.bam} \
              --GTF {input.gtf} \
              --TE {input.te} \
              --project counts/{wildcards.sample}
              """

rule align:
   input:
       fq1=lambda wildcards: expand(f"fastq/{config['samples'][wildcards.sample]}_L001_R1_001.fastq.gz"),
       fq2=lambda wildcards: expand(f"fastq/{config['samples'][wildcards.sample]}_L001_R2_001.fastq.gz"),
       genome=STAR_INDEX,
       gtf=GTFFILE
   output:"{sample}_Aligned.out.bam"
   log: "logs/{sample}_align_sort.log"
   threads: 16
   resources:
       mem_mb=36384,
       time='04:00:00'
   shell:"""
       STAR --genomeDir {input.genome} \
       --outFileNamePrefix {wildcards.sample}_ \
       --readFilesIn {input.fq1} {input.fq2} \
       --readFilesCommand zcat \
       --runThreadN {threads} \
       --genomeLoad NoSharedMemory \
       --outSAMattributes All \
       --outFilterMultimapNmax 100 \
       --winAnchorMultimapNmax 100 \
       --outSAMstrandField intronMotif \
       --outSAMtype BAM Unsorted \
       --sjdbGTFfile {input.gtf} 
       mkdir -p starlogs
       mv -f {wildcards.sample}_Log.final.out {wildcards.sample}_Log.out {wildcards.sample}_Log.progress.out {wildcards.sample}_SJ.out.tab {wildcards.sample}__STARgenome starlogs
       """

rule sort:
        input: "{sample}_Aligned.out.bam"
        output: "alignment/{sample}.sorted.bam"
        threads: 4
        resources: 
            mem='5G',
            time='30:00'
        shell:"""
                samtools sort -m 1G -@ {threads} -O bam -T {output}.tmp {input} -o {output};
                rm {wildcards.sample}_Aligned.out.bam
                """

```

# Load Libraries and Read in Count Matrix

Count files from alignment pipeline were then loaded into R for differential expression analysis using edgeR and downstream analysis of gene ontology.

```{r load}
library(biomaRt)
library(dplyr)
library(edgeR)
library(ggplot2)
library(tidyverse)


#Shifted to use TEcount so count matrices are output in parallel on the cluster, so we have a single count matrix for each sample. These count matrices need to be merged into one table.

counts <- lapply(list.files("counts", pattern = "*.cntTable", full.names = TRUE), read.table, header = TRUE) %>% reduce(left_join, by = "gene.TE") 



#rename columns to nice names
colnames(counts)<-gsub("alignment.","",colnames(counts))
colnames(counts)<-gsub(".sorted.bam","",colnames(counts))
colnames(counts)[1] <- "ensembl_gene_id"

counts <- counts[,-which(names(counts) %in% c("IgG_2","Iglon5_3" ))] #it was an outlier on Multidimensional scaling plot (MDS) & this animal had excessive gliosis and was removed
```

## Set Metadata

```{r metadata}
metadata = data.frame(row.names = c("IgG_1", # Animal 617
                                    #"IgG_2", # Animal 649 # this animal had excessive gliosis and was removed
                                    "IgG_3", # Animal 659
                                    "IgG_4", # Animal 667
                                    "Iglon5_1", # Animal 618
                                    "Iglon5_2", # Animal 650
                                    #"Iglon5_3", # Animal 660
                                    "PBS_1", # Animal 648
                                    "PBS_2", # Animal 658
                                    "PBS_3", # Animal 661
                                    "PBS_4" # Animal 666
                                    ), 
                      animal_id = c("617",
                                    #"649", # Animal 649 removed
                                    "659","667","618",
                                    "650",
                                    #"660",
                                    "648","658","661","666"),
                      group = c(rep("IgG", 3), rep("Iglon5", 2),  rep("PBS", 4)))


metadata$group = factor(metadata$group, levels = c("PBS", "IgG","Iglon5"))
group = metadata$group

```

## Annotate Count Matrix with Extra Genic Information

```{r annotate, tidy = TRUE, tidy.opts=list(width.cutoff=60)}
library(biomaRt)
library(dplyr)


#remove suffixes on ENSEMBL IDs if there
counts$ensembl_gene_id <- gsub(counts$ensembl_gene_id, pattern="\\.[0-9]+$", replacement="")

##biomaRt Annotations ( to add more attributes use listAttributes(mart) )
mart <- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))

#listAttributes(mart) # - to add more attributes
ID_QueryAttributes = c("ensembl_gene_id", "external_gene_name", "gene_biotype", "chromosome_name")
ID_QueryFilters = "ensembl_gene_id"
tmp1 <- data.frame("ensembl_gene_id" = counts$ensembl_gene_id)

#biomaRt get gene names
ids <- getBM(attributes = ID_QueryAttributes,
             filters = ID_QueryFilters,
             values = tmp1[, "ensembl_gene_id"], 
             mart = mart)

#Merge Gene Annotations and Rename symbol column containing ensembl ids to 'ensembl_gene_id'
colnames(counts)[1] <- "ensembl_gene_id"
counts = left_join(x = counts, y = ids, by = "ensembl_gene_id")


# sometimes BM returns duplicate rows, just make sure that there are no duplicates
dim(ids)
dim(unique(ids))


#Repeats aren't annotated in biomaRT so move them over to external_gene_name column
counts$external_gene_name[is.na(counts$external_gene_name)] <- as.character(counts$ensembl_gene_id[is.na(counts$external_gene_name)])
#Separate Rep familly
counts$rep_familly[is.na(counts$gene_biotype)] <- sub(".*:([-A-Za-z0-9]+):.*", "\\1", counts$ensembl_gene_id[is.na(counts$gene_biotype)], perl=TRUE)
#Separate Rep Name 
counts$rep_name[is.na(counts$gene_biotype)] <- sub("([A-Za-z0-9]+):.*", "\\1", counts$ensembl_gene_id[is.na(counts$gene_biotype)], perl=TRUE)
#Separate Rep Class
counts$rep_class[is.na(counts$gene_biotype)] <- gsub(".*:([A-Za-z0-9]+)", "\\1", counts$ensembl_gene_id[is.na(counts$gene_biotype)])
#Add Repeat Class to gene_biotype
counts$gene_biotype[is.na(counts$gene_biotype)] <- gsub(".*:", "", counts$ensembl_gene_id[is.na(counts$gene_biotype)])
#Add Rep Name to external gene name
counts$external_gene_name[!is.na(counts$rep_name)] <- as.character(counts$rep_name[!is.na(counts$rep_name)])

#move gene annotation columns to beginning
counts <- counts %>% dplyr::select(ensembl_gene_id, 
                            chromosome_name, 
                            external_gene_name, 
                            gene_biotype, 
                            rep_class, 
                            rep_familly, 
                            rep_name, 
                            everything()
                            )


```

# Analysis of Differential Gene Expression using EdgeR


```{r edgeR, results = 'asis', tidy = TRUE, tidy.opts=list(width.cutoff=60)}
library(edgeR)

#subset to protein coding
#counts <- subset(counts, gene_biotype == "protein_coding")

table.length <- ncol(counts)

#Put Count table in DGElist

dge <- DGEList(counts = counts[ , (length(ID_QueryAttributes) + 4) : table.length], 
               genes = counts[ , 1:(length(ID_QueryAttributes) + 3)], 
               group = group)


#Experimental Design
design = model.matrix(~0+group, data = dge$samples)
colnames(design) <- gsub("group", "", colnames(design))


#Set up contrasts
contr.matrix <- makeContrasts(
  Iglon5.vs.PBS = Iglon5 - PBS,
  Iglon5.vs.IgG = Iglon5 - IgG,
  levels = colnames(design))
contr.matrix


dge <- calcNormFactors(dge)#, method = "TMM")
dge <- estimateDisp(dge, design = design)
plotMDS(dge)
plotBCV(dge)

#Some ENSEMBL ids are gene duplicates, keep the ones with the highest counts.
o.counts <- order(rowSums(dge$counts), decreasing=TRUE)
dge.counts <- dge[o.counts,]
dups.counts <- duplicated(dge$genes$external_gene_name)
dge <- dge[!dups.counts,]
summary(dups.counts) 

#Remove lowly expressed Genes
keep <- rowSums(cpm(dge)>1) >= 2
dge <- dge[keep, , keep.lib.sizes=FALSE]

#Change rownames to gene name
rownames(dge$counts) <- dge$genes$external_gene_name

# fit general linear model for quasi-likelihood F test
fit <- glmQLFit(dge, design = design)
Iglon5.v.PBS.DE <- glmQLFTest(fit, contrast = contr.matrix[,"Iglon5.vs.PBS"]) 
Iglon5.v.IgG.DE <- glmQLFTest(fit, contrast = contr.matrix[,"Iglon5.vs.IgG"]) 

summary(decideTests(Iglon5.v.PBS.DE))


PVALUE = 0.05


Iglon5.PBS.DE.top <- topTags(Iglon5.v.PBS.DE, n=250, adjust.method="BH", sort.by = "logFC") 
Iglon5.PBS.DE.top.list <- rownames(rbind(Iglon5.PBS.DE.top$table))

PBS.df <- cbind(Iglon5.v.PBS.DE$table, Iglon5.v.PBS.DE$genes)
PBS.top.df <- Iglon5.PBS.DE.top$table

write.table(PBS.df, file = "Iglon5_differential_expression.csv", sep = ",", row.names = FALSE)
write.table(PBS.top.df, file = "Iglon5_differential_expression_top250.csv", sep = ",", row.names = FALSE)

Iglon5.IgG.DE.top <- topTags(Iglon5.v.IgG.DE, n=250, adjust.method="BH", sort.by = "logFC", p.value = PVALUE) 
Iglon5.IgG.DE.top.list <- rownames(rbind(Iglon5.IgG.DE.top$table))

IgG.df <- Iglon5.IgG.DE.top$table

logCPM <- cpm(dge, log=TRUE, prior.count = 1)

plotMDS(logCPM, col=as.numeric(group))

logCPM.DE <- as.data.frame(subset(logCPM, rownames(logCPM) %in% Iglon5.PBS.DE.top.list)) 
logCPM.DE.heat <- t(scale(t(logCPM.DE), scale=T)) 

```


```{r DE, results = 'none', tidy = TRUE, tidy.opts=list(width.cutoff=60)}

library(RColorBrewer)
library(gplots)
library(ggplot2)
library(ggrepel)

#RColorBrewer sequential palletes: 
#Blues BuGn BuPu GnBu Greens Greys Oranges OrRd PuBu PuBuGn PuRd Purples RdPu Reds YlGn YlGnBu YlOrBr YlOrRd

hmcol<-brewer.pal(9,"YlGnBu")
dev.set(dev.next()) # make sure graphics in valid state
dev.off() # make sure graphics in valid state
heat <- heatmap.2(logCPM.DE.heat, col = hmcol, tracecol = NA) 


volcano <- ggplot(PBS.df, aes(x = logFC, y = -log10(PValue),
                               color = ifelse(PBS.df$external_gene_name %in% rownames(PBS.top.df),
                                          PBS.df$gene_biotype, NA),
                              label=ifelse(
              PBS.df$external_gene_name %in% rownames(Iglon5.PBS.DE.top$table) &
              abs(PBS.df$logFC) > 2.5 &
             # PBS.df$logCPM > 0.5 &
              Iglon5.PBS.DE.top$table[as.character(PBS.df$external_gene_name), "FDR"] < 0.3,
              as.character(PBS.df$external_gene_name),
              ''))
                  )+
  geom_point(size = 0.5)+
  xlab("LogFC") +
  ylab("-log10 (p value)") +
  geom_text_repel() +
  theme_minimal() +
  theme(legend.title=element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle(expression(paste("Differential Gene Expression in Iglon5 Injected Cortices"))) 
volcano
ggsave("plots/Volcano_biotype.pdf", plot = last_plot(),width = 30, height = 15, units = "cm")

volcano <- ggplot(PBS.df, aes(x = logFC, y = -log10(PValue),
                               color = ifelse(PBS.df$external_gene_name %in% rownames(PBS.top.df),
                                          "red", NA),
                              label=ifelse(
              PBS.df$external_gene_name %in% rownames(Iglon5.PBS.DE.top$table) &
              abs(PBS.df$logFC) > 2.5 &
             # PBS.df$logCPM > 0.5 &
              Iglon5.PBS.DE.top$table[as.character(PBS.df$external_gene_name), "FDR"] < 0.3,
              as.character(PBS.df$external_gene_name),
              ''))
                  )+
  geom_point(size = 0.5)+
  xlab("LogFC") +
  ylab("-log10 (p value)") +
  geom_text_repel() +
  theme_minimal() +
  theme(legend.title=element_blank(), plot.title = element_text(hjust = 0.5)) +
  ggtitle(expression(paste("Differential Gene Expression in Iglon5 Injected Cortices"))) 
volcano
ggsave("plots/Volcano_sig.pdf", plot = last_plot(),width = 30, height = 15, units = "cm")

```
